src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1085 @@
+                                    .entry(batch_description)
+                                    .or_insert_with(Vec::new);
+
+                                batches.push((batch, batch_and_metadata.metrics));
+                            }
+
+                            continue;
+                        }
+                        Event::Progress(frontier) => {
+                            batches_frontier = frontier;
+                        }
+                    }
+                }
+                else => {
+                    // All inputs are exhausted, so we can shut down.
+                    return;
+                }
+            };
+
+            // Peel off any batches that are not beyond the frontier
+            // anymore.
+            //
+            // It is correct to consider batches that are not beyond the
+            // `batches_frontier` because it is held back by the writer
+            // operator as long as a) the `batch_description_frontier` did
+            // not advance and b) as long as the `desired_frontier` has not
+            // advanced to the `upper` of a given batch description.
+
+            let mut done_batches = in_flight_descriptions
+                .iter()
+                .filter(|(lower, _upper)| !PartialOrder::less_equal(&batches_frontier, lower))
+                .cloned()
+                .collect::<Vec<_>>();
+
+            trace!(
+                "persist_sink {collection_id}/{shard_id}: \
+                    append_batches: in_flight: {:?}, \
+                    done: {:?}, \
+                    batch_frontier: {:?}, \
+                    batch_description_frontier: {:?}",
+                in_flight_descriptions,
+                done_batches,
+                batches_frontier,
+                batch_description_frontier
+            );
+
+            // Append batches in order, to ensure that their `lower` and
+            // `upper` lign up.
+            done_batches.sort_by(|a, b| {
+                if PartialOrder::less_than(a, b) {
+                    Ordering::Less
+                } else if PartialOrder::less_than(b, a) {
+                    Ordering::Greater
+                } else {
+                    Ordering::Equal
+                }
+            });
+
+            for done_batch_metadata in done_batches.into_iter() {
+                in_flight_descriptions.remove(&done_batch_metadata);
+
+                let mut batches = in_flight_batches
+                    .remove(&done_batch_metadata)
+                    .unwrap_or_else(Vec::new);
+
+                trace!(
+                    "persist_sink {collection_id}/{shard_id}: \
+                        done batch: {:?}, {:?}",
+                    done_batch_metadata,
+                    batches
+                );
+
+                let (batch_lower, batch_upper) = done_batch_metadata;
+
+                let mut batch_metrics = BatchMetrics::default();
+                let mut to_append = batches
```

Comment:
am I missing something or could this just do a for loop over batches to `+=` the metrics?

Reply:
I need a vector of batches, without the batch metrics for the `compare_and _append`!